src/main/java/org/wildfly/security/audit/RotatingFileAuditEndpoint.java

```diff
@@ -0,0 +1,324 @@
+
+        final File file = getFile();
+        calcNextRollover(file != null && file.lastModified() > 0 ? file.lastModified() : System.currentTimeMillis());
+        if (rotateOnBoot && maxBackupIndex > 0 && file != null && file.exists() && file.length() > 0L) {
+            rotate(file);
+        }
+    }
+
+    @Override
+    protected void write(byte[] bytes) throws IOException {
+        super.write(bytes);
+        currentSize += bytes.length;
+    }
+
+    @Override
+    protected void preWrite(Date date) {
+        final long recordMillis = date.getTime();
+        if (recordMillis >= nextRollover) { // time based rollover
+            try {
+                final File file = getFile();
+                closeStreams(); // close the original file (some OSes won't let you move/rename a file that is open)
+                final Path target = Paths.get(file.getAbsolutePath() + nextSuffix);
+                Files.move(file.toPath(), target, StandardCopyOption.REPLACE_EXISTING);
+                setFile(file);
+                currentSize = 0;
+            } catch (IOException e) {
+                audit.unableToRotateLogFile(e);
+            }
+            calcNextRollover(recordMillis);
+        } else if (currentSize > rotateSize && maxBackupIndex > 0) { // file size based rollover
+            try {
+                final File file = getFile();
+                if (file == null) {
+                    // no file is set; a direct output stream or writer was specified
+                    return;
+                }
+                rotate(file);
+                currentSize = 0;
+            } catch (IOException e) {
+                audit.unableToRotateLogFile(e);
+            }
+        }
+    }
+
+    /**
+     * Moves file to file.1, file.1 to file.2 etc. Removes file.{maxBackupIndex}
+     */
+    private void rotate(final File file) throws IOException {
+        closeStreams();
+        final Path fileWithSuffix = Paths.get(file.getAbsolutePath() + nextSuffix);
+        Files.deleteIfExists(Paths.get(fileWithSuffix + "." + maxBackupIndex));
+        for (int i = maxBackupIndex - 1; i >= 1; i--) {
+            final Path src = Paths.get(fileWithSuffix + "." + i);
+            if (Files.exists(src)) {
+                final Path target = Paths.get(fileWithSuffix + "." + (i + 1));
+                Files.move(src, target, StandardCopyOption.REPLACE_EXISTING);
+            }
+        }
+        Files.move(file.toPath(), Paths.get(fileWithSuffix + ".1"), StandardCopyOption.REPLACE_EXISTING);
+        setFile(file);
+    }
+
+    /**
+     * For given time and period obtains time when should be new log file started
+     */
+    private void calcNextRollover(final long fromTime) {
+        if (period == Period.NEVER || format == null) {
+            nextRollover = Long.MAX_VALUE;
+            return;
+        }
+        nextSuffix = format.format(new Date(fromTime));
+        final Calendar calendar = Calendar.getInstance(timeZone);
```

Comment:
For the record, I think we should never, ever use Date or Calendar in any code that uses Java 8.  The java.time API is far superior.  We don't have to change this PR *but* if we run into any problems with this code, I'd strongly advise that we move over to java.time instead of messing around with Calendar any more.